syntax = "proto2";
package lchf;

message Row_f {
    repeated float value = 1;
}

message Mat_f {
    repeated Row_f row = 1;
}

message Row_i {
    repeated int32 value =1;
}

message Mat_i {
    repeated Row_i row =1;
}

message Mat_i_3 {
    repeated Mat_i channel =1;
}

message Info {
    required Mat_f rpy = 1;
    required Mat_f t = 2;
    required string id = 3;
}

message Infos {
    repeated Info info = 1;
}

message Linemod_embedding_ele {
    required int32 x = 1;
    required int32 y = 2;
    required int32 label = 3;
}

message Linemod_embedding_ele_vector {
    repeated Linemod_embedding_ele element = 1;
}

message Linemod_embedding {
    required Linemod_embedding_ele_vector rgb_embedding = 1;
    required Linemod_embedding_ele_vector depth_embedding = 2;
    repeated Mat_i rgb_response = 3;
    repeated Mat_i dep_response = 4;
    required int32 center_dep = 5;
}

message Linemod_feature {
    optional Mat_i_3 rgb = 1;
    optional Mat_i depth = 2;
    optional Mat_i mask = 3;
    optional Linemod_embedding embedding = 4;
    optional Info info = 5;
}

message Linemod_features {
    repeated Linemod_feature features = 1;
}

message Node {
    repeated int32 cnodes = 1;
    required int32 pnode = 2;
    required int32 depth = 3;
    required bool issplit = 4;
    repeated int32 ind_feats = 5;
    required int32 split_feat_idx = 6;
    required bool isleafnode = 7;
    optional float simi_thresh = 9;
}

message Tree {
    repeated Node nodes_ = 1;
    repeated int32 id_leafnodes_ = 2;
    required int32 max_depth_ = 3;
    required int32 num_leafnodes_ = 5;
    required int32 num_nodes_ = 6;
    required int32 size_thresh_ = 7;
    required int32 split_attempts_ = 8;
    required float simi_thresh_ = 9;
    repeated int32 id_non_leafnodes_ = 10;
}

message Forest {
    repeated Tree trees = 1;
    required int32 max_numtrees = 2;
    required double train_ratio = 3;
}





